Blindsight whip

Dariensg opened the pull request:
-Complete Implementation of Blindsight Whip Mechanics
-Neko will take care of the moving of the model when using the whip once the PR has been merged

Dariensg committed:
Blindsight Tongue Whip Beginnings

-Added Blindsight Tongue Whip
-Added recipe
-Added Blindsight Tongue Item Tier
-Several mechanics of the whip still need to be implemented (Check Project Page Card)

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into blindsight-whip

Dariensg committed:
Added Packets

-Created Packet Handler
-Created Basic Packet Superclass
-Created WhipReachPacket

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into blindsight-whip

Dariensg committed:
Whip Adjustments

-Added extended reach to the whip (doesn't work with cooldownAttackStrength yet)
-Added knockback to the whip
-Added ItemModelsProperties for whip to change textures (doesn't seem to be working)

Dariensg committed:
Merge pull request #42 from Dariensg/blindsight-whip

Blindsight Tongue Whip

caelwarner committed:
Whip now has variable reach

- Refactored packets to network
- Refactored IEPacketHandler to IENetworkHandler
- Switched packets to using a system with static methods
- Whip now had variable reach dependant on the attack cooldown
- Whip now changes texture when attacking

caelwarner committed:
Added blindsight tongue whip animation

- The animation takes 3 seconds which is roughly the attack cooldown

Dariensg committed:
Whip Changes

-Made whip charge up when holding right click
-Left clicking with whip will do damage, but won't knock entity back
-Model still needs to be adjusted, especially in third person
-Cooldown still needs to be removed on successful hit

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into blindsight-whip

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into blindsight-whip

Dariensg committed:
Removed Whip Cooldown

-When used properly by holding right click to charge, the whip no longer has a cooldown
-Fixed a mistake I made when merging with a previous commit in the mixins.json

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/InfernalExpansion.java:
```diff
@@ -94,18 +95,19 @@ private void commonSetup(final FMLCommonSetupEvent event) {
+		FlowerPotBlock flowerPot = (FlowerPotBlock) Blocks.FLOWER_POT;
+		flowerPot.addPlant(IEBlocks.DULLTHORNS.getId(), IEBlocks.POTTED_DULLTHORNS);
+		flowerPot.addPlant(IEBlocks.LUMINOUS_FUNGUS.getId(), IEBlocks.POTTED_LUMINOUS_FUNGUS);
+		flowerPot.addPlant(IEBlocks.SHROOMLIGHT_FUNGUS.getId(), IEBlocks.POTTED_SHROOMLIGHT_FUNGUS);
```
stop

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/client/InfernalExpansionClient.java:
```diff
@@ -19,13 +20,23 @@ public static void init() {
+			} else {
+				return (int) (((IWhipItem) itemStack.getItem()).getTicksSinceAttack() / 6.0F);
+			}
+		});
```
```suggestion
		ItemModelsProperties.registerProperty(IEItems.BLINDSIGHT_TONGUE_WHIP.get(), new ResourceLocation("attack_frame"), (itemStack, clientWorld, livingEntity) -> 
			livingEntity == null || livingEntity.getHeldItemMainhand() != itemStack ?
				0 : (int) (((IWhipItem) itemStack.getItem()).getTicksSinceAttack() / 6.0F)
		);
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/events/ClientEvents.java:
```diff
@@ -0,0 +1,56 @@
+
+		Vector3d eyePos = player.getEyePosition(1.0F);
+		Vector3d lookVec = player.getLookVec();
+		Vector3d reachVec = eyePos.add(lookVec.x * reach, lookVec.y * reach, lookVec.z * reach);
```
```suggestion
		Vector3d reachVec = eyePos.add(lookVec.mul(reach, reach, reach));
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/IWhipItem.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.UUID;
+
+public interface IWhipItem {
+	UUID REACH_MODIFIER_UUID = UUID.fromString("c6aac07c-3300-4994-be4a-8d42d171de2e");
```
Should be in a util class imo

Hellion-Games replied:
I like keeping it to the whip for now since it makes it easy to find and it's also the only thing that uses it. If we use it more in the future, I could see moving it to a util class, but currently being specific to the whip, I think it's fine where it is.

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/WhipItem.java:
```diff
@@ -0,0 +1,202 @@
+
+			Vector3d eyePos = playerentity.getEyePosition(1.0F);
+			Vector3d lookVec = playerentity.getLookVec();
+			Vector3d reachVec = eyePos.add(lookVec.x * reach, lookVec.y * reach, lookVec.z * reach);
```
```suggestion
			Vector3d reachVec = eyePos.add(lookVec.mul(reach, reach, reach));
```

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/WhipItem.java:
```diff
@@ -0,0 +1,202 @@
+		this.charging = true;
+	}
+
+    @Override
+	public void inventoryTick(ItemStack stack, World worldIn, Entity entityIn, int itemSlot, boolean isSelected) {
```
```suggestion
	@Override
	public void inventoryTick(ItemStack stack, World worldIn, Entity entityIn, int itemSlot, boolean isSelected) {
```
what the fuck

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/items/WhipItem.java:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	public float getDestroySpeed(ItemStack stack, BlockState state) {
+		Material material = state.getMaterial();
+		return material != Material.PLANTS && material != Material.TALL_PLANTS && material != Material.CORAL && !state.isIn(BlockTags.LEAVES) && material != Material.GOURD ? 1.0F : 1.5F;
+	}
+
+	public boolean canApplyAtEnchantingTable(ItemStack stack, Enchantment enchantment) {
+		return enchantment.type == (EnchantmentType.WEAPON);
+	}
+
+	public Multimap<Attribute, AttributeModifier> getAttributeModifiers(EquipmentSlotType equipmentSlot, ItemStack itemStack) {
+		ImmutableMultimap.Builder<Attribute, AttributeModifier> attributeBuilder = ImmutableMultimap.builder();
+		attributeBuilder.put(Attributes.ATTACK_DAMAGE, new AttributeModifier(ATTACK_DAMAGE_MODIFIER, "Weapon modifier", this.attackDamage, AttributeModifier.Operation.ADDITION));
+		attributeBuilder.put(Attributes.ATTACK_SPEED, new AttributeModifier(ATTACK_SPEED_MODIFIER, "Weapon modifier", this.attackSpeed, AttributeModifier.Operation.ADDITION));
+		attributeBuilder.put(ForgeMod.REACH_DISTANCE.get(), new AttributeModifier(REACH_MODIFIER_UUID, "Tool modifier", this.getReachDistanceModifier(), AttributeModifier.Operation.ADDITION));
+		Multimap<Attribute, AttributeModifier> attributes = attributeBuilder.build();
+		return equipmentSlot == EquipmentSlotType.MAINHAND ? attributes : super.getAttributeModifiers(equipmentSlot, itemStack);
+	}
+
+	@Override
+	public int getTicksSinceAttack() {
+		return ticksSinceAttack;
+	}
+
+	@Override
+	public boolean getAttacking() {
+		return attacking;
+	}
+
+	@Override
+	public boolean getCharging() {
+		return charging;
+	}
+
+	@Override
+	public void setAttacking(boolean value) {
+		attacking = value;
+	}
+}
```
please stay consistent if you use `this.` or not for properties

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/network/IENetworkHandler.java:
```diff
@@ -0,0 +1,39 @@
+		INSTANCE.messageBuilder(WhipReachPacket.class, index++).encoder(WhipReachPacket::encode).decoder(WhipReachPacket::decode).consumer(WhipReachPacket::handle).add();
+	}
+
+	public static <MSG> void sendToPlayer(MSG message, ServerPlayerEntity player)
+	{
+		INSTANCE.send(PacketDistributor.PLAYER.with(() -> player), message);
+	}
+
+	public static <MSG> void sendToAll(MSG message)
+	{
+		INSTANCE.send(PacketDistributor.ALL.noArg(), message);
+	}
+
+	public static <MSG> void sendToServer(MSG message)
+	{
+		INSTANCE.sendToServer(message);
+	}
```
```suggestion
	public static <MSG> void sendToPlayer(MSG message, ServerPlayerEntity player) {
		INSTANCE.send(PacketDistributor.PLAYER.with(() -> player), message);
	}

	public static <MSG> void sendToAll(MSG message) {
		INSTANCE.send(PacketDistributor.ALL.noArg(), message);
	}

	public static <MSG> void sendToServer(MSG message) {
		INSTANCE.sendToServer(message);
	}
```

SwanX1 reviewed (changes requested):
Minor changes needed here and there

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/client/InfernalExpansionClient.java

Co-authored-by: SwanX1 <karliscern@gmail.com>

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/events/ClientEvents.java

Co-authored-by: SwanX1 <karliscern@gmail.com>

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/items/WhipItem.java

Co-authored-by: SwanX1 <karliscern@gmail.com>

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/items/WhipItem.java

Co-authored-by: SwanX1 <karliscern@gmail.com>

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/network/IENetworkHandler.java

Co-authored-by: SwanX1 <karliscern@gmail.com>

Dariensg committed:
Changes for PR

-Changed tabs in main class to spaces (don't know how that happened to those lines)
-Made all calls to variables within WhipItem use "this" to be consistent

SwanX1 commented:
LGTM, merge at your discretion.
